### PR TITLE
Handle IForToLoopOperation over numeric types in control flow graph

### PR DESCRIPTION
**Pull request by AlekseyTs**

### CODE REVIEW

File: src/Compilers/Core/Portable/Operations/Loops/ForToLoopOperationInfo.cs
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/Loops/ForToLoopOperationInfo.cs]: Any reason this can't just be a struct?

File: src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs
Comment (333fred) [inline src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs]: >  [](start = 116, length = 1)

Lots of trailing whitespace.

File: src/Compilers/Core/Portable/SpecialTypeExtensions.cs
Comment (jaredpar) [inline src/Compilers/Core/Portable/SpecialTypeExtensions.cs]: Can we add a comment on the intent of this method or update the name. Unsure exactly what this is for at a glance.

File: src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs
Comment (jaredpar) [inline src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs]: >BaseForToLoopStatement [](start = 36, length = 22)

Which change is causing this generated file to get updated? Was expecting to see a change to on  of our XML files or generators but not seeing that.

Comment (mavasani) [reply src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs]: @jaredpar We have still not hooked up the generator to this file: https://github.com/dotnet/roslyn/issues/20204. We should probably rename this file to remove the `.generated.cs` suffix until that is implemented.

File: src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs
Comment (jaredpar) [inline src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs]: Consider adding a method to ForToLoopOperationUserDefinedInfo to force the evaluation of the lazies instances rather than enumerating the fields here. This would be more robust in the face of adding additional fields in the future.

File: src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs
Comment (jaredpar) [inline src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs]: >loopObject [](start = 26, length = 10)

The `loopObject` element could be a discard.

File: src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs
Comment (jaredpar) [reply src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs]: Ah okay. Thanks for the explanation.

As for renaming. I'm ambivalent. I don't want to create unnecessary work here if there is an issue tracking it. 

---
In reply to: [187120137](https://github.com/dotnet/roslyn/pull/26734#discussion_r187120137) [](ancestors = 187120137)

File: src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs
Comment (jaredpar) [inline src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs]: The `loopObject` element can be a discard.

File: src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs
Comment (jaredpar) [inline src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs]: This line looks suspicious. It implies there is a property here which is side effecting that needs to be executed.

File: src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs]: I'm planning on addressing the issue during the hackathon.

---
In reply to: [187121166](https://github.com/dotnet/roslyn/pull/26734#discussion_r187121166) [](ancestors = 187121166,187120137)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (jaredpar) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Why are invalid operations implicit?

File: src/Compilers/Core/Portable/Operations/Loops/ForToLoopOperationInfo.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/Loops/ForToLoopOperationInfo.cs]: > Any reason this can't just be a struct?

It is not common to need this information. The intent is to save some memory for the common case.

---
In reply to: [187115823](https://github.com/dotnet/roslyn/pull/26734#discussion_r187115823) [](ancestors = 187115823)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >getLoopControlVariableRefernce [](start = 70, length = 30)

typo: getLoopControlVariableRefernce -> getLoopControlVariableReference

File: src/Compilers/Core/Portable/SpecialTypeExtensions.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/SpecialTypeExtensions.cs]: > Can we add a comment on the intent of this method or update the name. Unsure exactly what this is for at a glance. 

Will do

---
In reply to: [187118574](https://github.com/dotnet/roslyn/pull/26734#discussion_r187118574) [](ancestors = 187118574)

File: src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs
Comment (AlekseyTs) [reply src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs]: > This line looks suspicious. It implies there is a property here which is side effecting that needs to be executed. 

The purpose of this line is to evaluate the property to make sure it doesn't crash. This is a test code.

---
In reply to: [187121790](https://github.com/dotnet/roslyn/pull/26734#discussion_r187121790) [](ancestors = 187121790)

File: src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs
Comment (AlekseyTs) [reply src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs]: > Consider adding a method to ForToLoopOperationUserDefinedInfo to force the evaluation of the lazies instances rather than enumerating the fields here. This would be more robust in the face of adding additional fields in the future. 

I do not expect us to add anything else to that class in the future. I also think this method would look strange in the class because it is only useful as a test.

---
In reply to: [187120366](https://github.com/dotnet/roslyn/pull/26734#discussion_r187120366) [](ancestors = 187120366)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: > Why are invalid operations implicit? 

Usually, implicit operations are those that are synthesized. Like this one, for example. It is not a clone of a node present in the original tree. 

---
In reply to: [187123049](https://github.com/dotnet/roslyn/pull/26734#discussion_r187123049) [](ancestors = 187123049)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (mavasani) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Can `StepValue` be null? If not, we probably want to update the [doc comment](http://source.roslyn.io/#Microsoft.CodeAnalysis/Operations/Loops/IForToLoopOperation.cs,36).

**Comment by 333fred** [comment]
Tests look good.

**Review by 333fred — APPROVED** [review]
Comments are OK to be addressed in a follow up PR.

**Comment by AlekseyTs** [comment]
@333fred, @mavasani, @jaredpar In my responses I identified the issues I am going to address in the next PR. These are:
- a comment on VBForToShiftBits https://github.com/dotnet/roslyn/pull/26734#pullrequestreview-118834516 
- a typo in getLoopControlVariableRefernce 
 https://github.com/dotnet/roslyn/pull/26734#pullrequestreview-118843522
- Reflect non-optional nature of StepValue value in doc comment and adjust TestOperationVisitor accordingly https://github.com/dotnet/roslyn/pull/26734#pullrequestreview-118848519

Please let me know if you consider this issues, or any other issues, blocking for this PR. Otherwise, could you please sign-off so that I can resolve merge conflicts and merge the PR, it is going to conflict with any other builder change merged into the branch.

**Review by mavasani — APPROVED** [review]

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler Please review, need a second review

**Review by jaredpar — APPROVED** [review]
:shipit:

**Pull request merged**
